Pybind11 integration

TanDro opened the pull request:
Improvements for the Labor at UMIT

- Improvments of plots
- Add Pybind11 functionality to pymoskito directly
- Bug fixing
- New example 'Tanksystem'

TanDro committed:
Improve plot feature to show color labels

TanDro committed:
Improve png export

TanDro committed:
Add example for cpp implementation, add pybind11 as lib, start implementation of cpp controller class

TanDro committed:
Add example for cpp implementation, add pybind11 as lib, start implementation of cpp controller class

TanDro committed:
Add export of datapoints directly from listwidget

TanDro committed:
Update example for cpp pid implementation

TanDro committed:
First implementation of Cpp module for controller with running of make

TanDro committed:
Update test scenario to 2 tanks, add first approach for observer, Add write and test of CMakeLists.txt

TanDro committed:
Update cpp integration to run under windows

TanDro committed:
Improve plot coloring

TanDro committed:
Add high gain observer to tanksystem example, update code a bit

TanDro committed:
Make Observer module compileable, can't return array

TanDro committed:
Add c++ version of observer for pymoskito with wrapper functions

TanDro committed:
Add exception handling at init of modules before simulation start

cklb commented:
Nice work!
Give me some time to have a look at your work so that we can incorporate this new feature.
Until then, I would like you think about to following points:
* Would it be possible to separate the Plotting/Color/Export fixes and improvements and the cpp backend into two PRs?
* Is the inclusion of pybind as submodule mandatory? (I would like to think of it as an optional dependency plus people seem to have their issues with submodules)
* Does this solution work or is there a way to get this to work if used from a pypy installation instead of a working copy?

Thank you for your work, I highly appreciate it.

TanDro committed:
Add colors to log dock

TanDro commented:
* Yes it should possible to seperate the features, give me some time :>
* Good point, it was also in my mind, but i don't have a good solution for this yet.
* I don't have check this till now, but will.

TanDro committed:
First possible solution for install with pybind11

TanDro commented:
I have tested the last commit 
(https://github.com/cklb/pymoskito/pull/74/commits/1faea06b23c1997a38f26074608681040fbb1946) to build a wheel and it runs without problems. I had seen in other projects, that simular approaches are used. The advantage if pymoskito have pybind as submodule, we can provide a well-functioning commit.

cklb commented:
>     * Yes it should possible to seperate the features, give me some time :>
Alright, I will focus on the pybind related lines then.

>     * Good point, it was also in my mind, but i don't have a good solution for this yet.
As it seems to me, for an installation of pybind11 via pip the line
`python -m pybind11 --includes`
yields the required headers for the  `g++` calls. Furthermore the way mentioned in https://github.com/pybind/python_example  seems like a good way to go.

cklb commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "pymoskito/libs/pybind11"]
```
As already stated, I would like to avoid this.

cklb commented on pymoskito/examples/tanksystem/binding/Controller.cpp:
```diff
@@ -0,0 +1,80 @@
+/** @file Controller.cpp
+ * This file includes the different controller implementations for the two tank system.
+ *
+ * Copyright (c) 2018 IACE
```
This may cause problems due different license for the rest of the project.

cklb commented on pymoskito/examples/tanksystem/binding/Controller.cpp:
```diff
@@ -0,0 +1,80 @@
+	double dDifferential = (dError - this->dLastError) / this->dSampleTime;
+
+	// Compute PID output
+    double dOut = 0.0;
```
tab vs. space?

cklb commented on pymoskito/examples/tanksystem/binding/Controller.cpp:
```diff
@@ -0,0 +1,80 @@
+}
+
+
+#endif // CONTROLLER_CPP
```
maybe add a newline?

cklb commented on pymoskito/examples/tanksystem/binding/Controller.h:
```diff
@@ -0,0 +1,129 @@
+
+#include <math.h>
+
+#define M_G 9.81
```
Left over from refactoring?

cklb commented on pymoskito/examples/tanksystem/binding/Observer.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+
+#if !defined(__AVR__)
```
This seems very specific, I think the focus of the example should be primarily on how to get the code running in pymoskito.

TanDro replied:
You are totally right.

cklb commented on pymoskito/examples/tanksystem/binding/Observer.cpp:
```diff
@@ -0,0 +1,64 @@
+{
+    double du = 0.0;
+
+    double da1 = this->dAS1 * sqrt(2 * M_G / (pow(this->dAT1, 2) - pow(this->dAS1, 2)));
```
How is the simulation of an internal model handled here?

TanDro replied:
There is done an euler integration. That's why the class needs the sample time.

cklb commented on pymoskito/examples/tanksystem/binding/binding_Observer.cpp:
```diff
@@ -0,0 +1,17 @@
+    .def("compute", &HighGainObserver::compute, "Calculates the observer output")
+    .def("setGain", &HighGainObserver::setGain)
+    .def("setInitialState", &HighGainObserver::setInitialState);
+}
```
Yet again, newline.

cklb commented on pymoskito/examples/tanksystem/controller.py:
```diff
@@ -0,0 +1,73 @@
+        self.lastU = 0
+        try:
+            from binding.Controller import PIDController
+            self.pid = PIDController()
```
You distinguish object creation and object initialization and yet there is an extra create call. Why is that?

cklb commented on pymoskito/examples/tanksystem/model.py:
```diff
@@ -0,0 +1,86 @@
+from . import settings as st
+
+
+class TwoTankSystem(pm.Model):
```
Maybe the example can be extended to use an external model? (Motivated by model reduction tools, ad export or similar)

cklb commented on pymoskito/examples/tanksystem/observer.py:
```diff
@@ -0,0 +1,128 @@
+
+        y = system_output[0]
+        uA = system_input[0]
+        res = self.obs.compute(y, uA)
```
I guess the need for an extra wrapper module arises from the inheritance issues? Certainly, a direct substitute would be most satisfying.

cklb commented on pymoskito/simulation_gui.py:
```diff
@@ -283,7 +284,13 @@ def __init__(self):
-        self.dataPointManipulationLayout.addStretch(0)
+        self.dataPointExportButton = QPushButton(chr(0x25BC), self)
+        self.dataPointExportButton.setToolTip(
+            "Export the selected data set from the left to the selected plot "
```
Export to the plot on the right?

cklb commented on pymoskito/simulation_gui.py:
```diff
@@ -589,6 +598,11 @@ def addDatapointToTree(self):
                 child = QTreeWidgetItem()
                 child.setText(1, dataPoint)
+
+                colorIdxItem = toplevelItem.childCount() % len(self.TABLEAU_COLORS)
```
Nice idea!

cklb commented on pymoskito/simulation_modules.py:
```diff
@@ -93,6 +98,103 @@ def calc_output(self, input_vector):
+        moduleName = settings['Module']
+
+        if os.name == 'nt':
+            bindingPath = os.getcwd() + '\\binding'
```
`os.sep.join()` ?

cklb replied:
This may render this distinction unnecessary.

cklb commented on pymoskito/simulation_modules.py:
```diff
@@ -93,6 +98,103 @@ def calc_output(self, input_vector):
+            if result.returncode == 0:
+                result = subprocess.run(['cmake', '--build', '.', '--config', 'Release'], cwd=bindingPath, shell=True)
+        else:
+            result = subprocess.run(['cmake . && make'], cwd=bindingPath, shell=True)
```
This means every time an CppSimulationModule is instantiated the cmake config is recreated and a build process is tarted?

TanDro replied:
No, cmake checks, if all needed files are avaiable and/or the config has changed. If nothing happens, cmake & make don't rebuild the libs.

cklb replied:
Ok, my fault. Better formulation: The build process is always started but will do nothing if all files are up to date.
I imagine that your use-case is to have the pymoskito gui up and running while making changes in the .cpp files. This way you may test your changes without restarting to gui.
From my point of view, it would be nice to bundle to whole cmake config generation into a helper function, offered by pymoskito which can be executed when the example is imported (e.g. in its __init__ or __main__ module) and therefore not re-generating the whole config every time an object is instantiated. 
The only remaining line in the constructor would then be the make call.
This way the object files are recreated when pymoskito is restartet

TanDro replied:
Okay I understand what you need and I'm with you. But I would to it at the registration of the module. I have renew the code for these. The idea is, that the user is unconcerned about the compile process, only about the code writing.

cklb commented on pymoskito/simulation_modules.py:
```diff
@@ -326,6 +434,12 @@ def _observe(self, time, system_input, system_output):
         pass
 
 
+class CppObserver(Observer, CppSimulationModule):
```
As stated above, a direct replacement for an observer object only makes sense if the compile logic can be moved into a helper method, thus rendering these mix-in classes unnecessary.

TanDro replied:
Yes you totally right. I have adaped the code for these.

TanDro committed:
Remove inheritance classes for cpp from SimualtionModule, Format controller and observer code, Remove copyright

TanDro commented:
>As it seems to me, for an installation of pybind11 via pip the line
python -m pybind11 --includes
yields the required headers for the g++ calls. Furthermore the way mentioned in https://github.com/pybind/python_example seems like a good way to go.

The main problem with this approach is, that the module is compiled with setuptools at the installation time. By the current solution, we want compile on change and we need next to the includes also the CMakeLists.txt from pybind for the `pybind11_add_module` command.

cklb commented on pymoskito/simulation_modules.py:
```diff
@@ -72,60 +72,23 @@ def __init__(self, settings):
-
-    def __init__(self, settings):
-        QObject.__init__(self, None)
+    def _generateBinding(self, settings):
```
This method only depends on `settings["Module"]` and therefore may be refactored as a function in the `tools` module.

TanDro committed:
Move generateBinding to tools, als helper function,
At init only the make process is started for the c code,
Remove binding keyword (not necessary)

cklb committed:
Some thoughts on the bindings:
* Refactored the src files to contain one class per file
* Renamed the Controller class to fit the name of its binding
  (Observer is still to be done)
* Moved the binding generation out of the registration process.
  The registry module should not have anything to do with something else
  than registration.
* The introduced extra property `Module` not quite meaningful and should
  not be the way to do things. The whole process should not depend on
  whether a pure python or a wrapper module is used.
* Refactored the binding generation in the `tools` module to get it pep8
  conform and to remove code duplicates.
* As it seems to me atm an installation of `pybind11` via pip misses the
  required `.cmake` files for the `find_package(pybind11)` call to
  succeed. I'm opening an issue upstream to get further information on
  this topic.
* After a little though on the topic, I really dislike the variant on
  using pybind as a git submodule since it greatly complicates the
  installation process of pymoskito for an optional dependency.
* Added a pybind entry to the requirements

cklb commented:
Please see the latest commit message for more feedback on the issue.
Furthermore, please remember to isolate the changes concerning pybind integration.

TanDro commented:
* Extract gui functions to pull request https://github.com/cklb/pymoskito/pull/75
* Extract pybind functions to pull request https://github.com/cklb/pymoskito/pull/76
